File: Inventory.cpp
#include "Inventory.hpp"

Inventory::Inventory() : plates(100), spoons(50), forks(50), knives(50), startingPlates(100), startingSpoons(50), startingForks(50), startingKnives(50) {
    // Initialize ingredients with default quantities
    ingredients["flour"] = 1000; 
    ingredients["sugar"] = 500; 
    ingredients["salt"] = 300; 
    ingredients["pepper"] = 200; 
    ingredients["oil"] = 1000; 
    ingredients["butter"] = 500; 
    ingredients["eggs"] = 100; 
    ingredients["milk"] = 1000; 
    ingredients["cheese"] = 300; 
}

void Inventory::addIngredient(const std::string& ingredient, int quantity) {
    ingredients[ingredient] += quantity;
}

std::string Inventory::useIngredient(const std::string& ingredient, int quantity) {
    std::string output;
    
    if (ingredients.find(ingredient) != ingredients.end()) {
        if (ingredients[ingredient] >= quantity) {
            ingredients[ingredient] -= quantity;
            output = "Ingredient " + ingredient + " used successfully!";
        } else {
            output = "Insufficient quantity of " + ingredient + "!";
        }
    } else {
        output = "Ingredient " + ingredient + " not found!";
    }
    
    return output;
}

std::string Inventory::usePlate() {
    std::string output;
    
    if (plates > 0) {
        plates--;
        output = "Plate used successfully!";
    } else {
        output = "No more plates available!";
    }
    
    return output;
}

std::string Inventory::useUtensils(int quantity) {
    std::string output;
    
    if (spoons >= quantity && forks >= quantity && knives >= quantity) {
        spoons -= quantity;
        forks -= quantity;
        knives -= quantity;
        output = "Utensils used successfully!";
    } else {
        output = "Insufficient utensils available!";
    }
    
    return output;
}

std::string Inventory::checkLowStock() {
    std::string output = "Checking low stock...\n";
    
    for (const auto& pair : ingredients) {
        if (pair.second < startingPlates / 5) {
            output += "Low stock warning: " + pair.first + " is running low!\n";
        }
    }
    
    if (plates < startingPlates / 5) {
        output += "Low stock warning: Plates are running low!\n";
    }
    
    if (spoons < startingSpoons / 5) {
        output += "Low stock warning: Spoons are running low!\n";
    }
    
    if (forks < startingForks / 5) {
        output += "Low stock warning: Forks are running low!\n";
    }
    
    if (knives < startingKnives / 5) {
        output += "Low stock warning: Knives are running low!\n";
    }
    
    return output;
}


std::string Inventory::checkDepleted() {
    std::string outputString = "Checking for depletion...\n";

    for (const auto& pair : ingredients) {
        if (pair.second == 0) {
            outputString += "Depletion warning: " + pair.first + " is depleted!\n";
        }
    }

    if (plates == 0) {
        outputString += "Depletion warning: Plates are depleted!\n";
    }
    if (spoons == 0) {
        outputString += "Depletion warning: Spoons are depleted!\n";
    }
    if (forks == 0) {
        outputString += "Depletion warning: Forks are depleted!\n";
    }
    if (knives == 0) {
        outputString += "Depletion warning: Knives are depleted!\n";
    }

    return outputString;
}

File: Inventory.js
//Written by: Chandler Finuf
//Tested by:
//Debugged by: Chandler Finuf, 
//const fetch = require('node-fetch');

export class Inventory {
    constructor() {
        // Initialize Inventory with empty items and ingredients, and base URL
        this.items = {};
        this.ingredients = {};
        this.baseUrl = 'https://torpid-closed-robe.glitch.me';
        // Fetch items and ingredients
        this.fetchItems();
        this.fetchIngredients();
    }

    // Fetch items from API and populate items object
    async fetchItems() {
        try {
            const response = await fetch(`${this.baseUrl}/items`);
            const items = await response.json();
            items.forEach(item => {
                this.items[item.name] = item.quantity;
            });
        } catch (error) {
            console.error('Error fetching items:', error);
        }
    }

    // Fetch ingredients from API and populate ingredients object
    async fetchIngredients() {
        try {
            const response = await fetch(`${this.baseUrl}/ingredients`);
            const ingredients = await response.json();
            ingredients.forEach(ingredient => {
                this.ingredients[ingredient.name] = ingredient.quantity;
            });
        } catch (error) {
            console.error('Error fetching ingredients:', error);
        }
    }

    // Add quantity to an ingredient
    addIngredient(ingredient, quantity) {
        this.ingredients[ingredient] += quantity;
    }

    // Use quantity of an ingredient
    useIngredient(ingredient, quantity) {
        if (this.ingredients[ingredient] >= quantity) {
            this.ingredients[ingredient] -= quantity;
            return `Ingredient ${ingredient} used successfully!`;
        } else {
            return `Insufficient quantity of ${ingredient}!`;
        }
    }

    // Use a plate
    usePlate() {
        if (this.items['plates'] > 0) {
            this.items['plates']--;
            return "Plate used successfully!";
        } else {
            return "No more plates available!";
        }
    }

    // Use utensils
    useUtensils(quantity) {
        if (
            this.items['spoons'] >= quantity &&
            this.items['forks'] >= quantity &&
            this.items['knives'] >= quantity
        ) {
            this.items['spoons'] -= quantity;
            this.items['forks'] -= quantity;
            this.items['knives'] -= quantity;
            return "Utensils used successfully!";
        } else {
            return "Insufficient utensils available!";
        }
    }

    // Check for low stock of ingredients and items
    checkLowStock() {
        let output = "Checking low stock...\n";
        for (const [ingredient, quantity] of Object.entries(this.ingredients)) {
            if (quantity < this.items['plates'] / 5) {
                output += `Low stock warning: ${ingredient} is running low!\n`;
            }
        }
        if (this.items['plates'] < this.items['plates'] / 5) {
            output += "Low stock warning: Plates are running low!\n";
        }
        if (this.items['spoons'] < this.items['spoons'] / 5) {
            output += "Low stock warning: Spoons are running low!\n";
        }
        if (this.items['forks'] < this.items['forks'] / 5) {
            output += "Low stock warning: Forks are running low!\n";
        }
        if (this.items['knives'] < this.items['knives'] / 5) {
            output += "Low stock warning: Knives are running low!\n";
        }
        return output;
    }

    // Check for depletion of ingredients and items
    checkDepleted() {
        let output = "Checking for depletion...\n";
        for (const [ingredient, quantity] of Object.entries(this.ingredients)) {
            if (quantity === 0) {
                output += `Depletion warning: ${ingredient} is depleted!\n`;
            }
        }
        if (this.items['plates'] === 0) {
            output += "Depletion warning: Plates are depleted!\n";
        }
        if (this.items['spoons'] === 0) {
            output += "Depletion warning: Spoons are depleted!\n";
        }
        if (this.items['forks'] === 0) {
            output += "Depletion warning: Forks are depleted!\n";
        }
        if (this.items['knives'] === 0) {
            output += "Depletion warning: Knives are depleted!\n";
        }
        return output;
    }

    // List current ingredient quantities
    listIngredients() {
        let output = "Current Ingredient Quantities:\n";
        for (const [ingredient, quantity] of Object.entries(this.ingredients)) {
            output += `${ingredient}: ${quantity}\n`;
        }
        return output;
    }

    // List current inventory
    listInventory() {
        let output = "Current Inventory:\n";
        output += this.listIngredients();
        output += `Plates: ${this.items['plates']}\n`;
        output += `Spoons: ${this.items['spoons']}\n`;
        output += `Forks: ${this.items['forks']}\n`;
        output += `Knives: ${this.items['knives']}\n`;
        return output;
    }
}

//module.exports = Inventory;
